src/Orleans.Transactions/State/TransactionalState.cs

```diff
@@ -441,33 +480,47 @@ private bool ValidateRead(long transactionId, TransactionalResourceVersion? read
             if (logRecord.Version != readVersion.Value)
             {
                 if(this.logger.IsEnabled(LogLevel.Debug))
-                    this.logger.LogDebug($"ValidateRead failed, because version is not same as recorded by state log record of the same transaction.");
+                    this.logger.LogDebug("ValidateRead failed, because version is not same as recorded by state log record of the same transaction. TransactionId: {0}, ReadVersion: {1}, logVersion {2}", transactionId, readVersion.Value, logRecord.Version);
                 return false;
             }
 
             return true;
         }
 
-        private async Task DoRecovery()
+        private async Task DoLoad()
         {
+            this.logger.LogDebug("DoLoad");
             // load inital state
             TransactionalStorageLoadResponse<TState> loadResponse = await this.storage.Load(StateName);
-            
+
             this.eTag = loadResponse.ETag;
             this.metadata = Metadata.FromString(loadResponse.Metadata);
+            this.commitedState = loadResponse.CommittedState;
             this.highestReadTransactionId = this.metadata.HighestReadTransactionId;
-            this.version = this.metadata.HighestVersion;
-            this.value = loadResponse.CommittedState;
-            this.log.Clear();
             foreach (PendingTransactionState<TState> pendingState in loadResponse.PendingStates)
             {
+                if (this.logger.IsEnabled(LogLevel.Debug))
+                    this.logger.LogDebug("Rebuilding log from storage for {0}", pendingState.SequenceId);
                 this.log[pendingState.SequenceId] = new LogRecord<TState>
                 {
                     NewVal = pendingState.State,
                     Version = (TransactionalResourceVersion.TryParse(pendingState.TransactionId, out TransactionalResourceVersion version)) ? version : default(TransactionalResourceVersion)
                 };
             }
 
+            DoRecovery();
+        }
+
+        private void DoRecovery()
+        {
+            this.logger.LogDebug("DoRecovery");
+            // load inital state
+            this.version = this.metadata.HighestVersion;
+            this.value = this.log.TryGetValue(this.version.TransactionId, out LogRecord<TState> record)
+                ? record.NewVal
+                : this.commitedState;
+            this.log.Clear();
+
```

Comment:
So when a grain is activated, we can set the current state to an unconfirmed value, which gives the transaction a chance to commit still. Is that the purpose of this change or is it something else?

Reply:
This has two purposes.
1) Use common logic to set current state for both save and load.
2) In case of initial load, set the active state to the correct state (it was previously always being set to stable state which is only correct if there are no pending states.
- will cleanup up comments, as they are no longer correct.
- will consider procedure name change, as this is not really recovering.

Reply:
I think w.r.t naming, calling this "Recover" is fine, since it's common in fault tolerant systems to have a recovery process which runs on startup to bring the system to a consistent state.